### PR TITLE
Improve algorithm for loading Java classes from Slice type IDs

### PR DESCRIPTION
**Pull request by pepone**

### CODE REVIEW

File: java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java [hunk 1]
```diff
@@ -720,90 +720,82 @@ public Class<?> apply(String typeId)
...
 
         //
-        // It's a new type ID, so first convert it into a Java class name.
+        // The fully qualified class name was already resolved, keep using it.
         //
-        if(fullyQualifiedClassName == null)
+        if (fullyQualifiedClassName != null)
         {
-            fullyQualifiedClassName = com.zeroc.Ice.Util.typeIdToClass(typeId);
-            addClass = true;
```
Comment (pepone) [inline java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java:749]: `addClass` is no longer required as we return earlier if this is a class we have previously resolved.

File: java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java [hunk 2]
```diff
@@ -720,90 +720,82 @@ public Class<?> apply(String typeId)
...
+        // If we didn't find the class yet, try the default package prefix or without prefix.
+        if (c == null)
         {
-            String pkg = _initData.properties.getProperty("Ice.Default.Package");
-            if(pkg.length() > 0)
-            {
-                c = getConcreteClass(pkg + "." + fullyQualifiedClassName);
-            }
+            String packagePrefix = _initData.properties.getProperty("Ice.Default.Package");
+            c = getConcreteClass(
```
Comment (pepone) [inline java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java:786]: load the class by first trying to prepend the default package prefix to the fully qualified name returned by typeIdToClass. If no default package prefix is defined (i.e., packagePrefix is empty), we then directly use the fully qualified name as returned by typeIdToClass

File: java/test/src/main/java/test/Ice/classLoader/AllTests.java [hunk 3]
```diff
@@ -134,7 +134,6 @@ public static void allTests(test.TestHelper helper, boolean collocated)
                     out.flush();
                     ConcreteClass cc = initial.getConcreteClass();
                     test(cc != null);
-                    test(classLoader.check("Test.ConcreteClass"));
```
Comment (pepone) [inline java/test/src/main/java/test/Ice/classLoader/AllTests.java:137]: This corresponded to the failed first attempt without the default package prefix.

File: java/test/src/main/java/test/Ice/classLoader/AllTests.java [hunk 4]
```diff
@@ -153,7 +152,6 @@ public static void allTests(test.TestHelper helper, boolean collocated)
                 catch(E ex)
                 {
                 }
-                test(classLoader.check("Test.E"));
```
Comment (pepone) [inline java/test/src/main/java/test/Ice/classLoader/AllTests.java:156]: ditto

File: java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java [hunk 5]
```diff
@@ -720,90 +720,82 @@ public Class<?> apply(String typeId)
...
+        // To convert a Slice type ID into a Java class, the following steps are taken:
         //
-        // 1. Convert the Slice type id into a classname (e.g., ::M::X -> M.X).
-        // 2. If that fails, extract the top-level module (if any) from the type id
-        //    and check for a Package property. If found, prepend the property
-        //    value to the classname.
-        // 3. If that fails, check for an Default.Package property. If found,
-        //    prepend the property value to the classname.
+        // 1. Convert the Slice type ID into a classname (e.g., ::M::X becomes M.X).
+        // 2. Check if the application has defined any package prefixes for the top-level module (e.gm "M").
```
Comment (bernardnormier) [inline java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java:726]: The e.g. looks incorrect.

File: java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java [hunk 6]
```diff
@@ -720,90 +720,82 @@ public Class<?> apply(String typeId)
...
+        //    Attempt to resolve the <package-prefix>+<class-name> in order trying each defined package
+        //    prefix.
+        // 3. If the above step fails, search for a Default.Package property. If found,
```
Comment (bernardnormier) [inline java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java:729]: The property is Ice.Default.Package. Why "a" Default.Package property?

File: java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java [hunk 7]
```diff
@@ -720,90 +720,82 @@ public Class<?> apply(String typeId)
...
+        //    prepend its value to the classname. Otherwise, attempt to use the
+        //    classname directly.
```
Comment (bernardnormier) [inline java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java:731]: Just to confirm: Ice.Package.TopLevelModuleName is now a list but Ice.Default.Package remains a single value?

File: java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java [hunk 8]
```diff
@@ -720,90 +720,82 @@ public Class<?> apply(String typeId)
...
+            String[] packagePrefixes = _builtInModulePackagePrefixes.get(topLevelModule);
+            if (packagePrefixes == null)
             {
-                String topLevelModule = typeId.substring(2, pos);
-                String[] packagePrefixes = _builtInModulePackagePrefixes.get(topLevelModule);
-                if (packagePrefixes == null)
-                {
-                    packagePrefixes = _initData.properties.getPropertyAsListWithDefault("Ice.Package." + topLevelModule, null);
-                }
+                packagePrefixes = _initData.properties.getPropertyAsListWithDefault("Ice.Package." + topLevelModule, null);
```
Comment (bernardnormier) [inline java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java:766]: Why WithDefault? The implicit default is presumably null.

**Review by bernardnormier — APPROVED** [review]

File: java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java (hunk 7 above)
Comment (pepone) [reply java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java:731]: Yes, that is correct.

File: java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java (hunk 8 above)
Comment (pepone) [reply java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java:766]: You are right, getPropertyAsList defaults to null.

**Pull request merged**